src/attach-shadow.js

```diff
@@ -237,25 +293,141 @@ ShadyRoot.prototype._updateChildNodes = function(container, children) {
     for (let j=s.index, n; j < s.index + s.addedCount; j++) {
       n = children[j];
       insertBefore.call(container, n, next);
-      // TODO(sorvell): is this splice strictly needed?
       composed.splice(j, 0, n);
     }
   }
 }
 
-ShadyRoot.prototype.getInsertionPointTag = function() {
-  return this._distributor.insertionPointTag;
+/**
+ * Adds the given slots. Slots are maintained in an dom-ordered list.
+ * In addition a map of name to slot is updated.
+ */
+ShadyRoot.prototype._addSlots = function(slots) {
+  let slotNamesToSort;
+  for (let i=0; i < slots.length; i++) {
+    let slot = slots[i];
+    // ensure insertionPoints's and their parents have logical dom info.
+    // save logical tree info
+    // a. for shadyRoot
+    // b. for insertion points (fallback)
+    // c. for parents of insertion points
+    slot.__shady = slot.__shady || {};
+    recordChildNodes(slot);
+    recordChildNodes(slot.parentNode);
+    let name = this._nameForSlot(slot);
+    if (this._slotMap && this._slotMap[name]) {
+      slotNamesToSort = slotNamesToSort || {};
+      slotNamesToSort[name] = true;
+    }
+    this._slotList.push(slot);
+  }
+  if (slotNamesToSort) {
+    for (let n in slotNamesToSort) {
+      let slotsForName = this._extractSlotsOfName(n);
+      let sorted = this._sortSlots(slotsForName);
+      this._slotList.push(...sorted);
+    }
+  }
+  this._updateSlotMap();
+}
+
+ShadyRoot.prototype._nameForSlot = function(slot) {
+  return slot['name'] || slot.getAttribute('name') || CATCHALL_NAME;
+}
+
+ShadyRoot.prototype._extractSlotsOfName = function(name) {
+  let slots;
+  for (let i=0; i < this._slotList.length; i++) {
+    let slot = this._slotList[i];
+    let n = this._nameForSlot(slot);
+    if (n == name) {
+      this._slotList.splice(i, 1);
+      i--;
+      slots = slots || [];
+      slots.push(slot);
+    }
+  }
+  return slots;
+}
+
+/**
+ * Slots are kept in an ordered list. Here they are sorted by dom-tree order.
+ */
+ShadyRoot.prototype._sortSlots = function(slots) {
+  return slots.sort((a, b) => {
```

Comment:
I was about to say, "just use [`Node.compareDocumentPosition`](https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition)!".  But I remember probably talking about that and that it obviously won't work for logical positions.

OTOH you've got most of `compareDocumentPosition` implemented here, so might bookmark that as something to actually polyfill and then just use it here.